ports/stm32/usbd_cdc_interface.c

```diff
@@ -121,13 +125,10 @@ int8_t usbd_cdc_control(usbd_cdc_state_t *cdc_in, uint8_t cmd, uint8_t *pbuf, ui
             break;
 
         case CDC_SET_LINE_CODING:
-            #if 0
-            LineCoding.bitrate = (uint32_t)(pbuf[0] | (pbuf[1] << 8) | \
-                (pbuf[2] << 16) | (pbuf[3] << 24));
-            LineCoding.format = pbuf[4];
-            LineCoding.paritytype = pbuf[5];
-            LineCoding.datatype = pbuf[6];
-            /* Set the new configuration */
+            #if MICROPY_HW_USB_CDC_1200BPS_TOUCH
+            if (*((uint32_t *)pbuf) == 1200) {
+                powerctrl_enter_bootloader(0, 0);
```

Comment:
This will enter the bootloader immediately, ie the USB won't have a chance to respond back to the host and the host will see an abrupt loss of USB device.  Would be good to find a way to delay the entering of the bootloader, until after the USB has responded to this request.

Also, to support all the different ways of entering the bootloader, maybe better to call `machine_bootloader_obj.fun.var(0, NULL)` (or expose `machine_bootloader()` and call that directly).

Reply:
> or expose machine_bootloader() and call that directly

Yes I can add that in a commit

> and the host will see an abrupt loss of USB device. Would be good to find a way to delay the entering of the bootloader, until after the USB has responded to this request.

But that's okay for USB devices, normally I just disconnect boards. Otherwise, how about this, a fix both issues above, what if I schedule machine_bootloader with a static scheduler node ?

Reply:
The problem is that, eg, if you write a small Python script that uses pyserial to do the 1200 touch, then it may hang for a bit and then raise an exception if the USB disappears before it responds.  At the low-level of any USB driver there will be some code which waits for a response, with a timeout.  That timeout will be triggered if the USB device never responds.

> what if I schedule machine_bootloader with a static scheduler node

Yes I was thinking something along those lines.  That should work.

Reply:
Is this for port `stm32` or all ports ?

Reply:
All other ports will use TinyUSB, and I'm not sure how the line-state callback works there.  Is it called synchronously, or delayed?  It's probably synchronous so would also need to schedule a call to the bootloader.

I know this scheduling will add code size (and some complexity).  So maybe it's worth checking first to see if the host PC really does have problems if the USB device resets before responding.

Also, how does it work on boards running Arduino?

Reply:
>  So maybe it's worth checking first to see if the host PC really does have problems if the USB device resets before responding.

I would say no because USB device can be disconnected at any time, and because it's been used for a very long time, but at the same time don't mind delaying the call (it does add extra code that might not be necessary)

Reply:
```
STATIC NORETURN mp_obj_t machine_bootloader(size_t n_args, const mp_obj_t *args) {
    #if MICROPY_HW_ENABLE_USB
    pyb_usb_dev_deinit();
 ```

Wouldn't this provide a somewhat clean disconnect ?

Reply:
> I would say no because USB device can be disconnected at any time

But not usually during a USB transfer.  Usually you disconnect USB while it's idle.

See eg ea81bcf1c0ddeb8be34192cbededecfa425c039b which improves mboot DFU exit behaviour (`pydfu.py -x`) so that it's not an abrupt loss of USB device.  That commit definitely improved usability of `pydfu.py` with mboot.

> Wouldn't this provide a somewhat clean disconnect?

No, because the USB driver is not reentrant (it would be calling into the USB driver while it's in a USB IRQ).

Reply:
Just for review, please see the last commit, will squash later if approved.

Reply:
Yes that looks good.